moveit_setup_assistant/moveit_setup_controllers/src/ros2_controllers_config.cpp

```diff
@@ -0,0 +1,224 @@
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: David V. Lu!! */
+
+#include <moveit_setup_controllers/ros2_controllers_config.hpp>
+
+namespace moveit_setup
+{
+namespace controllers
+{
+void ROS2ControllersConfig::onInit()
+{
+  control_xacro_config_ = config_data_->get<ControlXacroConfig>("control_xacro");
+}
+
+void ROS2ControllersConfig::loadPrevious(const std::filesystem::path& package_path, const YAML::Node& /*node*/)
+{
+  changed_ = false;
+  controllers_.clear();
+
+  // Load moveit controllers yaml file if available-----------------------------------------------
+  std::filesystem::path ros_controllers_yaml_path = package_path / CONTROLLERS_YAML;
+  std::ifstream input_stream(ros_controllers_yaml_path);
+  if (!input_stream.good())
+  {
+    RCLCPP_WARN_STREAM((*logger_), "Does not exist " << ros_controllers_yaml_path);
+    return;
+  }
+
+  // Begin parsing
+  try
+  {
+    YAML::Node config = YAML::Load(input_stream);
+
+    // Parse Controller Manager Config
+    const YAML::Node& cm_node = config["controller_manager"]["ros__parameters"];
+    if (!cm_node.IsDefined())
+    {
+      return;
+    }
+    for (const auto& it : cm_node)
+    {
+      if (!it.second.IsMap())
+        continue;
+
+      const auto& type_node = it.second["type"];
+      if (!type_node.IsDefined())
+      {
+        continue;
+      }
+
+      ControllerInfo controller;
+      controller.name_ = it.first.as<std::string>();
+      controller.type_ = type_node.as<std::string>();
+
+      if (controller.type_ != "joint_state_broadcaster/JointStateBroadcaster")
+      {
+        controllers_.push_back(controller);
+      }
+    }
+
+    // ParseIndividual Controller Configs
+    for (auto& controller : controllers_)
+    {
+      const YAML::Node& controller_node = config[controller.name_]["ros__parameters"];
+      if (!controller_node.IsDefined() || !controller_node.IsMap())
+      {
+        continue;
+      }
+      auto jnode = controller_node["joints"];
+      if (jnode.IsDefined() && jnode.IsSequence())
+      {
+        // Looping through sequences with yaml is sometimes buggy when using iterators
+        // so here we use a "classic" loop and disable the clang-tidy rule
+        for (std::size_t i = 0; i < jnode.size(); i++)  // NOLINT(modernize-loop-convert)
+        {
+          controller.joints_.push_back(jnode[i].as<std::string>());
+        }
+      }
+
+      if (controller.joints_.empty())
+      {
```

Comment:
Is this basically creating one default joint, if none are defined? Prob should print a warning in this case.

```suggestion
      if (controller.joints_.empty())
      {
        RCLCPP_WARN_STREAM_ONCE((*logger_), "Did not find even one valid joint in the ros2_control yaml `joints` list. Creating a dummy default.");
```